### DISCUSSION TITLE
Icon-Label Mapping for Another Tag Visual Display

### DISCUSSION THREAD

**Discussion by guoranjun2** [freeplane/freeplane#2008] [Ideas]
Hi there,

First off, thanks for all the hard work on tag feature! 

Here is my idea, images can sometimes convey more information than text. Would it be possible to add a feature where, if a label matches a file name in a designated icon/image folder, the label would be displayed as an icon instead of text? I think this could really improve the visual experience.

Just a thought—would love to hear what you think!

**Comment by macmarrum** [comment]
Hi @guoranjun2 

If by **label** you mean node (core) text, it's already possible, using scripting (aka macros).

Your icon files (.svg, .png, .jpg) need to be placed in `<user-dir>/icons/`.

You need to download [MacmarrumNodeChangeListenerForTextToIcon.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/lib/MacmarrumNodeChangeListenerForTextToIcon.groovy) and save it in `<user-dir>/lib/`, then restart Freeplane.

In your mind map, you need to create a scrip `Tools->Edit script…` and paste the following code:
```groovy
node.mindMap.addListener(new MacmarrumNodeChangeListenerForTextToIcon())
```
then run it `Actions->Run`.

From now on, any change of node text will trigger **MacmarrumNodeChangeListenerForTextToIcon**, which will look in your user icons for a match (icon name without its extension) and if found insert the corresponding icon and remove the node text.

_**Note**: To find your `<user-dir>` you can go to `Tools->Open user directory`_

---

The script `node.mindMap.addListener(new MacmarrumNodeChangeListenerForTextToIcon())` must be run each time the mind map is opened. There is a way to automate it, by using the add-on https://github.com/macmarrum/freeplane-ScriptExecutionOnMindmapOpen. Download it, open and agree to install it, then in your mind map's root node create an attribute named `scriptOnMapOpen` with the value `node.mindMap.addListener(new MacmarrumNodeChangeListenerForTextToIcon())`.

---

**MacmarrumNodeChangeListenerForTextToIcon** can search for icons also in other places, like details, note, attributes.

**Reply by guoranjun2** [reply]
Thank you for your thoughtful response. Your script is quite interesting, but I believe there might be a slight misunderstanding. My intention was to have the tag's style rendered as an image (when the corresponding image file is present), to better utilize the new feature.
I think it’s just an interesting idea that hasn’t been implemented yet. Imagine if we used people’s names as tags and rendered them as images using the corresponding local files—wouldn’t that make it more intuitive?

**Reply by macmarrum** [reply]
OK, I see now what you mean.

I know that tags were requested by the Freeplane community as a means to add textual _icons_ to nodes.

Before tags were introduces, some people created SVG icons containing text and used them as a tags substitute.

I'm not sure I'd want to mix images and text in tags. If I want an image, I use an icon. If I want text, I use a tag.

That being said, it's possible to automatically add/remove an icon representing a tag when a tag is added/removed, with the help of a NodeChangeListener. So when I add a tag named `macmarrum`, if there's a corresponding file available (`<user-dir>/icons/macmarrum.svg`), a `macmarrum` icon is added automatically. When I remove the tag `macmarrum`, the `macmarrum` icon is also removed.

**Reply by guoranjun2** [reply]
> OK, I see now what you mean.
> 
> I know that tags were requested by the Freeplane community as a means to add textual _icons_ to nodes.
> 
> Before tags were introduces, some people created SVG icons containing text and used them as a tags substitute.
> 
> I'm not sure I'd want to mix images and text in tags. If I want an image, I use an icon. If I want text, I use a tag.
> 
> That being said, it's possible to automatically add/remove an icon representing a tag when a tag is added/removed, with the help of a NodeChangeListener. So when I add a tag named `macmarrum`, if there's a corresponding file available (`<user-dir>/icons/macmarrum.svg`), a `macmarrum` icon is added automatically. When I remove the tag `macmarrum`, the `macmarrum` icon is also removed.

What you said makes sense, I'll try using the icons.
